nimbus/src/dbcache.rs

```diff
@@ -29,10 +29,18 @@ pub struct DatabaseCache {
 
 impl DatabaseCache {
     // Call this function whenever it's possible that anything cached by this
-    // struct (eg, our enrollments) might have changed. It is passed a
-    // &Database, which implies some mutex guarding that Database is held.
-    pub fn update(&self, db: &Database) -> Result<()> {
-        let experiments = get_enrollments(&db)?;
+    // struct (eg, our enrollments) might have changed.
+    //
+    // This function must be passed a `&Database` and a `Writer`, which it
+    // will commit before updating the in-memory cache. This is a slightly weird
+    // API but it helps encorce two important properties:
+    //
+    //  * By requiring a `Writer`, we ensure mutual exclusion of other db writers
+    //    and thus prevent the possibility of caching stale data.
+    //  * By taking ownership of the `Writer`, we ensure that the calling code
+    //    updates the cache after all of its writes have been performed.
+    pub fn commit_and_update(&self, db: &Database, writer: Writer) -> Result<()> {
+        let experiments = get_enrollments(&db, &writer)?;
```

Comment:
I'm not sure I understand yet: the writer hasn't committed; we're getting enrollments. Is this copy of the old data or the not yet committed data? I can see it's the same `writer`: so confused :(

Reply:
The idea is that by taking the `writer`, we read the state of the database *from the point-of-view of that transaction*, meaning we read and cache the changes written by the caller.

Reply:
(I added a comment to clarify)

Reply:
Ok, this makes sense here. 

Can I generalize the rule, by replacing (in my head) a getter with a writer is said to be a `precommit_getter`?

Reply:
That sounds right to me, yes - if you see a `get_*` function taking a `Writer` or `&Writer` then it's doing so in order to read the uncommitted state of the transaction.

(This particular case is a bit special in taking *ownership* of the `Writer` rather than a reference, so it's also taking the responsibility for committing the transaction; that doesn't really generalize and I would expect any other instances of this pattern to be taking a `&Writer` instead)